Concept usage

CrustaShrimp opened the pull request:
-Add user concepts for easier use later
-Use concepts directly instead of typename T requires concept<T>
-Add extra checks on iterators (InputIt being an std::input_interator, OutputIt being an std::output_iterator of correct type)

Solves a part of #1, the OutputIt type being unspecified and unconstrained.

CrustaShrimp committed:
Add custom concepts, use concepts directly instead of via requires

CrustaShrimp committed:
Additional requirements on input iterators, output iterators

PG1003 commented on src/golomb.h:
```diff
@@ -96,40 +97,39 @@ class encoder
-        , k( k )
-        , base( 1u << k )
-        , zeros_threshold( 1 + k )
+    encoder( OutputIt output_in, size_t k_in = 0u ) noexcept
```
Although the _in postfix is explicit it is not necessary.
I do prefer to keep the names as it was since the parameters contain the initialization values of the encoder object.

CrustaShrimp replied:
Today I learned... Initialiser list doesn't conflict with input parameters as far as names are concerned.
I was worried about the ambiguity of the names, since the members have the exact same name as the parameters.

PG1003 commented on src/golomb.h:
```diff
@@ -96,40 +97,39 @@ class encoder
+        static constexpr auto input_mask    = std::numeric_limits< UnsignedInputValueT >::max();
 
-        const auto overflow_threshold = std::numeric_limits< UnsignedInputValueT >::max() - ( static_cast< OutputDataT >( base ) );
+        const auto overflow_threshold = input_mask - ( static_cast< OutputDataT >( base ) );
```
Sharp eye. :-)
I see that the braces around the cast can be omitted too.

PG1003 commented on src/golomb.h:
```diff
@@ -96,40 +97,39 @@ class encoder
         const int  bit_width = static_cast< int >( std::bit_width( value ) ); // cast is a workaround for unresolved defect report in GCC/libc++
         const int  width     = overflow ? input_digits : bit_width;
-        const int  zeros     = overflow ? max_zeros : width - zeros_threshold;
+        const int  zeros     = overflow ? max_zeros : (width - zeros_threshold);
```
The braces are not needed here.
I'm a bit reluctant to use braces 'just to be sure'. A temporary value is more preferable but i think even that is not needed here.

CrustaShrimp replied:
Personal preference, I like to clarify this kind of stuff for future me, when debugging these steps are often uncertain. Was the purpose a turnary and then substracting zeros_threshold from the turnary result, or is one of the turnary results with - zeros_threshold?
In this case, it is the latter and the brackets are not required.

PG1003 commented on src/golomb.h:
```diff
@@ -33,11 +33,16 @@
+
+template<typename InputIt>
+concept unsigned_integral_iterator = std::unsigned_integral< typename std::iterator_traits< InputIt >::value_type >;
+
```
These these concepts should be moved into the detail namespace since they are not part of the API.

CrustaShrimp replied:
Done, in MSVC at first it required the std:: to be specified as global namespace (::std::), but it turned out to be a messup.
Code now compiles in MSVC and G++ (required extra change in the file, see other comment)

PG1003 reviewed (changes requested):
The use of concepts within 'template<   >' instead of the 'requires' is indeed cleaner.
There are a few minor changes I would like to see. One functional change but mostly to fit my undocumented coding style. :-)

CrustaShrimp committed:
Fix remarks except moving custom concepts into detail namespace

CrustaShrimp committed:
Move custom concepts into detail namespace.
This requires specifying std to be from the global namespace, only in those concepts.

CrustaShrimp committed:
Nevermind, the global namespace was not required after all.
G++ requires the scan_zeros at the declartion of the state member of class decoder to be specified as scan_state

CrustaShrimp commented on src/golomb.h:
```diff
@@ -224,7 +223,7 @@ class decoder
     int                  digits;
     
-    enum { scan_zeros, decode } state = scan_zeros;
+    enum class scan_state { scan_zeros, decode } state = scan_state::scan_zeros;
```
G++ 12.2.0 with -std=c++20 required the scan_zeros to be specified as scan_state::

CrustaShrimp replied:
Other issues have also been dealt with...

CrustaShrimp committed:
Create compiler specific concepts, since CLang and GCC don't support std::back_insert_iterators as output iterators and has extra restrictions on input iterators.

CrustaShrimp commented:
Nasty problems in GCC & CLang:
-The custom interators fail the requirements (std::input_iterator and std::output_iterator) in these compilers.
-Manually specifying the output_iterator requirement of type_correctness (the output iterator should contain the same type as the datatype) is also hard in case std::back_insert_iterators are used.
The underlaying type of a std::back_insert_iterators is unknown, so you cannot verify the type.

In MSVC, this seems to work fine.
In [latest commit](https://github.com/PG1003/golomb/pull/2/commits/7a3f9d99ef6b9c8ef6bb0e3048f67aecffa5ee18), a fix has been done using an ifdef (not pretty but it works) to check whether the compiler is MSVC. If so, allow the default concepts, otherwise use std::input_or_output_iterator for the input iterators and nothing for the output iterators.

CrustaShrimp committed:
-Revert concepts back to use std concepts for all compilers
-include _input_ in the names of the custom input iterator concepts
-Fix issue with  example code: dereference operator has to return alias type reference.
Note that the const version gives a warning about the const before reference being unnecessary, but it reads easier for the human compiler.
-Make binary_output_file_iterator movable by making its member non-const.

CrustaShrimp commented on src/golomb.h:
```diff
@@ -35,12 +35,11 @@ namespace pg::golomb
-
 template<typename InputIt>
-concept integral_iterator = std::integral< typename std::iterator_traits< InputIt >::value_type > && std::input_iterator<InputIt>;
+concept integral_input_iterator = std::integral< typename std::iterator_traits< InputIt >::value_type > && std::input_iterator< InputIt >;
```
renamed to include _input_

CrustaShrimp commented on util/golomb.cpp:
```diff
@@ -137,8 +137,8 @@ struct binary_input_file_iterator
 
-    T &                          operator*()        { return value; }
-    const T &                    operator*()  const { return value; }
+    reference                    operator*()        { return value; }
```
Did the trick, as you mentioned.

CrustaShrimp commented on util/golomb.cpp:
```diff
@@ -137,8 +137,8 @@ struct binary_input_file_iterator
-    T &                          operator*()        { return value; }
-    const T &                    operator*()  const { return value; }
+    reference                    operator*()        { return value; }
+    const reference              operator*()  const { return value; }
```
GCC in /Wall gives a warning that the first const is ignored, since it is automatically a const reference as a result of the member function being const. It is still more clear to the human "compiler" this way IMO

PG1003 replied:
I don't get any warning when building with the makefile. The CPPFLAGS include -Wall.

CrustaShrimp commented on util/golomb.cpp:
```diff
@@ -213,7 +218,7 @@ struct binary_output_file_iterator
 
 private:
-    std::FILE * const file = nullptr;
+    std::FILE * file = nullptr;
```
class is required to be movable, by the concept std::output_iterator<...>

CrustaShrimp commented on util/golomb.cpp:
```diff
@@ -181,6 +181,11 @@ struct binary_output_file_iterator
 
+    binary_output_file_iterator(const binary_output_file_iterator& other)                   = default; // Copy ctor
+    binary_output_file_iterator& operator= (const binary_output_file_iterator& other)       = default; // Copy assignment
+    binary_output_file_iterator(binary_output_file_iterator&& other)             noexcept   = default; // Move ctor
```
Required to be movable, these four could be removed but it does make it more clear IMO, although they can be removed since the issue is resolved by making the member non-const

PG1003 replied:
The comments are not needed here. I think its quite obvious what is going on here.
Also I'm in favor to remove _not_ add defaults when they are not needed.

CrustaShrimp replied:
Removed

CrustaShrimp reviewed (commented):
See comments

PG1003 commented on src/golomb.h:
```diff
@@ -304,17 +301,16 @@ class decoder
-template< typename OutputValueT,
-          typename InputIt,
+template< std::integral OutputValueT,
+          pg::golomb::detail::unsigned_integral_input_iterator InputIt,
```
`pg::golomb` can be removed since the template resides in this namespace.

CrustaShrimp replied:
Done

CrustaShrimp committed:
Remove unnecessary namespace usage, remove default copy and move ctors and assignments

CrustaShrimp committed:
Forgot a save

CrustaShrimp committed:
Attempt to comply with coding standards

CrustaShrimp commented:
Should now be up to date, made an attempt to comply with the coding standards and fixed the remaining issues.

PG1003 commented:
I've squashed and merged the changes to the main branch.
